File: examples/cpp/user_dictionary.cpp
/**
 * User dictionary - load custom vocabulary for domain-specific text
 *
 * Demonstrates: user dictionary loading from file and from memory,
 * before/after comparison of analysis results.
 */
#include <iostream>

#include "suzume.h"

void printMorphemes(const std::vector<suzume::core::Morpheme>& morphemes) {
  for (const auto& m : morphemes) {
    std::cout << "  " << m.surface << " ["
              << suzume::core::posToString(m.pos) << "]";
    if (!m.lemma.empty() && m.lemma != m.surface) {
      std::cout << " (lemma: " << m.lemma << ")";
    }
    std::cout << "\n";
  }
}

int main() {
  suzume::SuzumeOptions opts;
  opts.skip_user_dictionary = true;  // Start clean
  suzume::Suzume analyzer(opts);

  std::string text = "初音ミクのコスプレ衣装を作りました";

  // Without user dictionary
  std::cout << "Without user dictionary:\n";
  printMorphemes(analyzer.analyze(text));

  // Load user dictionary from memory (TSV format: surface\tPOS\treading)
  std::string dict_data = "初音ミク\tNOUN\tはつねみく\n";
  if (analyzer.loadUserDictionaryFromMemory(dict_data.data(),
                                            dict_data.size())) {
    std::cout << "\nWith user dictionary:\n";
    printMorphemes(analyzer.analyze(text));
  }

  // Or load from file:
  // analyzer.loadUserDictionary("/path/to/user.tsv");

  return 0;
}

File: examples/ts/user_dictionary.ts
/**
 * User dictionary - load custom vocabulary for domain-specific text
 *
 * Demonstrates: loading user dictionary from string data,
 * before/after comparison, and binary dictionary loading.
 *
 * Run: npx tsx examples/ts/user_dictionary.ts
 */
import { Suzume } from '../../dist/index.js';

const suzume = await Suzume.create();

const text = '初音ミクのコスプレ衣装を作りました';

// Without user dictionary
console.log('Without user dictionary:');
for (const m of suzume.analyze(text)) {
  console.log(`  ${m.surface} [${m.pos}]`);
}

// Load user dictionary (TSV format: surface\tPOS\treading)
const dictData = '初音ミク\tNOUN\tはつねみく\n';
const loaded = suzume.loadUserDictionary(dictData);
console.log(`\nDictionary loaded: ${loaded}`);

console.log('\nWith user dictionary:');
for (const m of suzume.analyze(text)) {
  const lemma = m.baseForm !== m.surface ? ` (lemma: ${m.baseForm})` : '';
  console.log(`  ${m.surface} [${m.pos}]${lemma}`);
}

// Binary dictionary: load pre-compiled .dic file
// const response = await fetch('/path/to/user.dic');
// const dicData = new Uint8Array(await response.arrayBuffer());
// suzume.loadBinaryDictionary(dicData);
